locust/dispatch.py

```diff
@@ -49,548 +48,133 @@ class UsersDispatcher(Iterator):
-        """
-        crash_dump = json.dumps(
-            {
-                "spawn_rate": self._spawn_rate,
-                "initial_dispatched_users": self._initial_dispatched_users,
-                "desired_users_assigned_to_workers": self._desired_users_assigned_to_workers,
-                "user_classes_count": self._user_classes_count,
-                "initial_user_count": sum(map(sum, map(dict.values, self._initial_dispatched_users.values()))),
-                "desired_user_count": sum(self._user_classes_count.values()),
-                "number_of_workers": self._number_of_workers,
-            },
-            indent="  ",
-        )
-        fp = tempfile.NamedTemporaryFile(
-            prefix="locust-dispatcher-crash-dump-", suffix=".json", mode="wt", delete=False
-        )
-        try:
-            fp.write(crash_dump)
-        finally:
-            fp.close()
-        return crash_dump, fp.name
+        self._dispatch_iteration_durations.append(delta)
 
-    @property
-    def _number_of_workers(self) -> int:
-        return len(self._users_left_to_assigned)
+        # print("Dispatch cycle took {:.3f}ms".format(delta * 1000))
 
-    @property
-    def _all_users_have_been_dispatched(self) -> bool:
-        return self._user_count_left_to_assigned == 0
+        if self._current_user_count == self._target_user_count:
+            # No sleep when this is the last dispatch iteration
+            return
 
-    @property
-    def _user_count_left_to_assigned(self) -> int:
-        return sum(map(sum, map(dict.values, self._users_left_to_assigned.values())))
+        sleep_duration = max(0.0, self._wait_between_dispatch - delta)
+        gevent.sleep(sleep_duration)
 
-    def _try_next_user_class_in_order_to_stay_balanced_during_ramp_up(self, user_class_to_add: str) -> bool:
-        """
-        Whether to skip to the next user class or not. This is done so that
-        the distribution of user class stays approximately balanced from one dispatch
-        iteration to another.
-        """
-        # For performance reasons, we use `functools.lru_cache()` on the `self._dispatched_user_classes_count`
-        # method because its value does not change within the scope of the current method. However, the next time
-        # `self._try_next_user_class_in_order_to_stay_balanced_during_ramp_up` is invoked, we need
-        # `self._dispatched_user_classes_count` to be recomputed.
-        self._dispatched_user_classes_count.cache_clear()
-
-        if all(user_count > 0 for user_count in self._dispatched_user_classes_count().values()):
-            # We're here because each user class have at least one user running. Thus,
-            # we need to ensure that the distribution of users corresponds to the weights.
-            if not self._adding_this_user_class_respects_distribution_better_than_adding_any_other_user_class(
-                user_class_to_add
+    def _ramp_up(self) -> Dict[str, Dict[str, int]]:
```

Comment:
the way I'm understanding this method is that it adds on user? Maybe rename it to indicate that?

Reply:
Something like `_add_users_on_workers` and `_remove_users_on_workers`?

Reply:
yes that sounds better!